doc/cmis-module-enhancement/cmis-module-enhancement.md

```diff
@@ -0,0 +1,224 @@
+To do that, Ports OA shall use the query capability for the following SAI attributes:
+
+1.	 SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE – to know whether the platform supports the control of enabling/disabling the high-speed signal from ASIC to module from SONIC.
+2.	SAI_SWITCH_ATTR_PORT_HOST_TX_READY_NOTIFY – to know whether the platform supports the asynchronous notification from SDK/SAI to SWSS about start/stop of transmission of high-speed signal data from ASIC to module.
+
+Ports OA shall use the status of support of these capabilities for its flow.
+For example, the notification consumer for a new SAI_SWITCH_ATTR_PORT_HOST_TX_READY_NOTIFY notification will be initialized only if the platform supports the asynchronous notification from SAI/SDK to SWSS.
+
+#### 6.3.2. Enabling/disabling “host_tx_signal”
+As described earlier, for platforms supporting this capability, high-speed signal shall be explicitly enabled/disabled by Ports OA on getting the Insertion/Removal event for a specific module.
+
+When a module is plugged in/out the PMON adds/deletes a per-port entry in the TRANSCIEVER_TABLE in STATE DB.
+This event shall be used by Ports OA as a trigger for enabling or disabling the host_tx_signal (using setting true/false to SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE on the Port Object).
+
+To detect this insertion/removal event the Ports Orch Agent shall register for changes in the TRANSCIEVER_TABLE in STATE DB and will do the following:
+* INSERTION event – set SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE to TRUE
+* REMOVAL event   - set SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE to FALSE
+
+Please note that on platforms supporting this functionality, SDK/FW shall start the transmission of high-speed signal from ASIC to a module only if both criteria below are met:
+* TX signal is enabled (via setting SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE to TRUE)
+* Admin Status is set to UP
+
+Whenever one of them is set to false (e.g. TX signal is disabled due to removal event of Admin station is set to DOWN), the SDK/FW stops transmission of this signal.
+
+This approach allows to avoid the issue of unterminated transmission when ASIC starts the transmission of the high-speed signal even when no module is plugged-in causing cross-talks to adjacent ports, high EMI/BER and shortening the transceiver lifespan
+
+On port creation for platforms supporting proposed enhancements, the SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE shall be explicitly set by Ports OA to FALSE since its default SAI value is defined as TRUE for backward-compatibility reasons.
+
+#### 6.3.3. Handling of HW-based “host_tx_ready” event
+Per description above, on the platform supporting new enhancements, the SDK/FW shall send an asynchronous notification of “host_tx_ready” to SWSS when the ASIC starts/stops sending high-speed signal to a module.
+
+The vendor SDK shall internally support this indication from ASIC (via the trap mechanism) and shall call the notification callback registered by Ports OA on initialization .
```

Comment:
How does SONiC be notified the whethernew  tx_trap_signal trap supported or not? I believe it must be vendor CMIS implemetation must support this trap otherwise correct us!

Reply:
SONIC (Ports OA) will perform the query capability for support of HW-based host_tx_ready notification. Then it will initializes the consumer for this notification only if Capability Query returned the indication of support on the platform. The vendor SAI will indicate its support of this notification ONLY if the trap from ASIC about start/stop of host_tx_ready signal is supported by ASIC/FW on the platform